### PR TITLE
🐛 Fix generic tuple cannot have more than one item

### LINKED ISSUE #6383: Wrong validation error when using `NamedTuple` with Generics in Model
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

Hi, 

I'm encountering a strange validation error when I use a `NamedTuple` combined with a Generic inside a Pydantic BaseModel.

When I run the attached example code, it leads to the following validation error:

`pydantic_core._pydantic_core.ValidationError: 1 validation error for A
p_value
  Tuple should have at most 1 item after validation, not 2 [type=too_long, input_value=(1, True), input_type=tuple]`

I have searched the documentation and found no reason that this should not be possible. However, when I remove the type parameter `[float]` from the declaration of `p_value` in `class A`, it runs without any validation errors. Unfortunately, this results in loss of validation information for the `value` field of `ProtectableValue`.

### Example Code

```Python
from typing import Generic, TypeVar, NamedTuple
from pydantic import BaseModel

T = TypeVar('T')

class ProtectableValue(NamedTuple, Generic[T]):
    value: T
    is_protected: bool

class A(BaseModel):
    p_value: ProtectableValue[float]


A.model_validate({'p_value': (1,True)})
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0
        pydantic-core version: 2.0.1 release build profile
               python version: 3.11.3 (main, Jun  5 2023, 09:32:32) [GCC 13.1.1 20230429]
                     platform: Linux-6.4.1-arch1-1-x86_64-with-glibc2.37
     optional deps. installed: ['email-validator', 'typing-extensions']
```

Selected Assignee: @samuelcolvin

### PR DESCRIPTION
**Pull request by lig**
## Change Summary

For now reproduces the bug with generic tuple cannot have more than one item 

## Related issue number

Fixes #6383 

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable
* [ ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @Kludex

### CODE REVIEW

**Comment by lig** [comment]
please review

File: tests/test_json_schema.py [hunk 1]
```diff
@@ -629,13 +629,11 @@ class Model(BaseModel):
...
         ),
-        (Tuple[str], {'prefixItems': [{'type': 'string'}], 'minItems': 1, 'maxItems': 1}),
-        (Tuple[()], {'maxItems': 0, 'minItems': 0}),
-        (
-            Tuple[str, ...],
-            {'items': {'type': 'string'}, 'type': 'array'},
-        ),
+        pytest.param(Tuple[str], {'prefixItems': [{'type': 'string'}], 'minItems': 1, 'maxItems': 1}, id='Tuple[str]'),
+        pytest.param(Tuple[()], {'maxItems': 0, 'minItems': 0}, id='Tuple[()]'),
+        pytest.param(Tuple[str, ...], {'items': {'type': 'string'}, 'type': 'array'}, id='Tuple[str, ...]'),
```
Comment (lig) [inline tests/test_json_schema.py:636]: I've added `pytest.param` with ids as one of those was failing and this way it's easier to see which one fails.

File: tests/test_generics.py [hunk 2]
```diff
@@ -2446,3 +2454,27 @@ class VariadicToolbox(BaseModel, Generic[ComponentVar, Unpack[NumberOfComponents
     with pytest.raises(TypeError):
         # Should raise error because integer 5 does not meet the bound requirements of ComponentVar
         MyToolboxClass(manual_toolset=(sa, sb, 5))
+
+
+@pytest.mark.skipif(
```
Comment (adriangb) [inline tests/test_generics.py:2459]: Does it work with typing extensions?

Comment (dmontagu) [reply tests/test_generics.py:2459]: I added support for recursive namedtuple in the latest commit on this branch, and as part of that, added a test that used `typing_extensions.NamedTuple` to get generics support in older versions of python. It's working for all older versions, so I'm pretty sure it's good.

(The change I made was fairly small, essentially just copied the contextmanager that was used in the typeddict_schema to handle recursion)

File: pydantic/_internal/_generate_schema.py [hunk 3]
```diff
@@ -841,20 +843,36 @@ def _typed_dict_schema(self, typed_dict_cls: Any, origin: Any) -> core_schema.Co
...
-    def _namedtuple_schema(self, namedtuple_cls: Any) -> core_schema.CallSchema:
+    def _namedtuple_schema(self, namedtuple_cls: Any, origin: Any) -> core_schema.CoreSchema:
         """Generate schema for a NamedTuple."""
-        annotations: dict[str, Any] = get_type_hints_infer_globalns(
-            namedtuple_cls, include_extras=True, localns=self.types_namespace
-        )
-        if not annotations:
-            # annotations is empty, happens if namedtuple_cls defined via collections.namedtuple(...)
-            annotations = {k: Any for k in namedtuple_cls._fields}
+        with self.defs.get_schema_or_ref(namedtuple_cls) as (namedtuple_ref, maybe_schema):
```
Comment (lig) [inline pydantic/_internal/_generate_schema.py:848]: @dmontagu I think, next time we touch this, we should think about generalizing this with the dataclass schema. The code looks fairly similar.

**Pull request merged**